app/src/main/java/org/liberty/android/fantastischmemo/ui/QuizActivity.java

```diff
@@ -116,25 +125,81 @@ public void onPostInit() {
     }
 
     @Override
-    public void onCreate(Bundle savedInstanceState){
+    public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         Bundle extras = getIntent().getExtras();
         categoryId = extras.getInt(EXTRA_CATEGORY_ID, -1);
         startCardOrd = extras.getInt(EXTRA_START_CARD_ORD, -1);
         quizSize = extras.getInt(EXTRA_QUIZ_SIZE, -1);
         shuffleCards = extras.getBoolean(EXTRA_SHUFFLE_CARDS, false);
+        timerMode = extras.getBoolean(EXTRA_TIMER_MODE, false);
+        timeInSeconds = extras.getInt(EXTRA_COUNTDOWN, 120);
+        timeLeftInMilliseconds = timeInSeconds * 1000;
+        countdownText = (TextView) findViewById(R.id.countdown_text);
+
         if (savedInstanceState != null) {
             startCardId = savedInstanceState.getInt(EXTRA_START_CARD_ID, -1);
         }
 
         getMultipleLoaderManager().registerLoaderCallbacks(3, new QuizQueueManagerLoaderCallbacks(), false);
 
         startInit();
+        if (timerMode) { startTimer(); }
+
+    }
+
+    public void startTimer() {
+        this.countDownTimer = new CountDownTimer(timeLeftInMilliseconds, 1000) {
+
+            @Override
+            public void onTick(long millisUntilFinished) {
+                timeLeftInMilliseconds = millisUntilFinished;
+                updateTimerText();
+            }
+
+            @Override
+            public void onFinish() {
+                if (getCurrentCard().getOrdinal() != totalQuizSize) {
+                    new AlertDialog.Builder(QuizActivity.this)
+                            .setTitle(R.string.quiz_not_completed)
+                            .setMessage("Would you like to try again?")
+                            .setPositiveButton(R.string.yes_text, flushAndQuitListener)
+                            .setCancelable(false)
+                            .show();
+                }
+            }
+        }.start();
+
+        timerMode = true;
+    }
+
+    public void stopTimer() {
+        if (countDownTimer != null) {
+            countDownTimer.cancel();
+            timerMode = false;
+        }
+    }
+
+    public void updateTimerText() {
+        int minutesLeft = (int) timeLeftInMilliseconds / 60000;
+        int secondsLeft = (int) timeLeftInMilliseconds % 60000 / 1000;
```

Comment:
Why did you put % 60000/1000 and not just % 60?

Reply:
because im using timeLeftInMilliseconds i gotta use 60000 to find the remainder and then convert by 1000 so it's in seconds.